File: apollosensors.py
from numpy import float32
import random
import openlocationcode as olc
from parameters import OLC_PRECISION, BATTERY_VOLTAGE


# Sensor-related functions (to be replaced with actual sensor code)

def get_location() -> (str, float32, float32):
    """
    Gets the location of the balloon
    :return: (str, np.float32, np.float32) - Open Location Code, latitude, longitude
    """
    # The location is an Open Location Code with a length of OLC_PRECISION significant digits.
    try:
        # This is a placeholder function. TODO: Replace with actual GPS code.
        lat = random.uniform(-90, 90)
        long = random.uniform(-180, 180)
    except (Exception,):
        lat = 0
        long = 0
    return olc.encode(lat, long, OLC_PRECISION + 9), float32(lat), float32(long)


def get_altitude() -> float32:
    """
    Gets the altitude of the balloon
    :return: float32 - Altitude, in meters
    """
    try:
        # This is a placeholder function. TODO: Replace with actual altitude code.
        altitude = float32(random.uniform(0, 10000))
    except (Exception,):
        altitude = float32(0.0)
    return altitude


def get_voltage() -> float32:
    """
    Gets the voltage difference from nominal of the balloon
    :return: float32 - Delta volts from nominal, in volts
    """
    try:
        # This is a placeholder function. TODO: Replace with actual voltage code.
        voltage = float32(BATTERY_VOLTAGE - random.uniform(0, 12))
    except (Exception,):
        voltage = float32(100.0)
    return voltage


def get_temperature() -> float32:
    """
    Gets the temperature of the balloon
    :return: float32 - Temperature, in degrees Celsius
    """
    try:
        # This is a placeholder function. TODO: Replace with actual temperature code.
        temperature = float32(random.uniform(-50, 50))
    except (Exception,):
        temperature = float32(0.0)
    return temperature

File: src/balloon-firmware/src/apollosensors.rs
extern crate open_location_code;
extern crate rand;

use alloc::string::String;
use rand::distributions::uniform::Uniform;
use rand::prelude::Distribution;
use rand::SeedableRng;
use rand::rngs::SmallRng;

use crate::parameters::OLC_CODE_LENGTH;

// Sensor-related functions (to be replaced with actual sensor code)

pub fn get_location() -> ([u8; OLC_CODE_LENGTH], [u8; 4], [u8; 4]) {

    
    // Returns a tuple containing the Open Location Code, latitude, and longitude
    // of the balloon.
    // TODO: Replace with actual sensor code.
    let mut rng = SmallRng::from_entropy();

    // Create a uniform distribution between -90 and 90 degrees.
    let latitude_range: Uniform<f32> = Uniform::new(-90.0, 90.0);

    // Create a uniform distribution between -180 and 180 degrees.
    let longitude_range: Uniform<f32> = Uniform::new(-180.0, 180.0);

    // Generate a random latitude and longitude.
    let latitude: f32 = latitude_range.sample(&mut rng);
    let longitude: f32 = longitude_range.sample(&mut rng);

    let point = geo_types::Point::new(latitude as f64, longitude as f64);
    // Populate the Open Location Code string.
    let mut olc_code: String = open_location_code::encode(point, OLC_CODE_LENGTH);

    let olc_code_str: &str = olc_code.as_mut_str();
    let mut olc_code_bytes: [u8; OLC_CODE_LENGTH] = [0; OLC_CODE_LENGTH];
    olc_code_bytes.copy_from_slice(olc_code_str.as_bytes());
    let latitude_bytes = latitude.to_be_bytes();
    let longitude_bytes = longitude.to_be_bytes();
    return (olc_code_bytes, latitude_bytes, longitude_bytes);
}

pub fn get_altitude() -> [u8; 4] {
    // Returns the altitude of the balloon, in meters.
    // TODO: Replace with actual sensor code.

    // Generate a random altitude between 0 and 10,000 meters.
    let mut rng = rand::thread_rng();
    let altitude_range = Uniform::new(0.0, 10000.0);
    let altitude: f32 = altitude_range.sample(&mut rng);
    return altitude.to_be_bytes();
}

pub fn get_voltage() -> [u8; 4] {
    // Returns the voltage difference from nomal voltage, in volts.
    // TODO: Replace with actual sensor code.

    // Generate a random voltage difference between -0.1 and 1.5 volts.
    let mut rng = rand::thread_rng();
    let voltage_range = Uniform::new(-0.1, 1.5);
    let voltage: f32 = voltage_range.sample(&mut rng);
    return voltage.to_be_bytes();
}

pub fn get_temperature() -> [u8; 4] {
    // Returns the temperature of the balloon, in degrees Celsius.
    // TODO: Replace with actual sensor code.

    // Generate a random temperature between -50 and 50 degrees Celsius.
    let mut rng = rand::thread_rng();
    let temperature_range = Uniform::new(-50.0, 50.0);
    let temperature: f32 = temperature_range.sample(&mut rng);
    return temperature.to_be_bytes();
}